src/sdk/service/network-validation-service/network-validation-service.ts

```diff
@@ -0,0 +1,475 @@
+   */
+  isValidPort(port: number): boolean {
+    return (port >= 1 && port <= 65535);
+  }
+
+  /**
+   * Gets the number of IP addresses in a given range. Although unlikely to be
+   * an issue, note that due to the fact that javascript numbers are represented
+   * as signed, two's complement integers, for a network with more than 2^31
+   * addresses this function will return a negative number (May not even be
+   * technically possible). Returns -1 if the parameters are invalid.
+   * @param {Range} ipRange object representing IP range of the form { start:
+   *          "x.x.x.x", end: "x.x.x.x" }
+   * @return {integer} size of range.
+   */
+  getNumberOfAddressesInRange(range: Range): number | null {
+    if (!(this.isValid(range.start) && this.isValid(range.end))) {
+      return null;
+    }
+    return this.getIpAsInteger(range.end) - this.getIpAsInteger(range.start) + 1;
+  }
+
+  /**
+   * Returns an available IP address within one of an array of given ranges,
+   * given an array of the IP addresses that are already allocated. Returns null
+   * if no addresses are unallocated.
+   * @param {Range[]} ipRanges: array of IP ranges to choose from.
+   * @param {string[]} allocatedIpAddresses: array of IP addresses which have
+   *          already been allocated.
+   * @return {string} an available IP address.
+   */
+  getAvailableIp(ipRanges: Range[], allocatedIpAddresses: string[]): string | undefined {
+    const allocInts = allocatedIpAddresses.map(
+      (ipAddress: string) => {
+        return this.getIpAsInteger(ipAddress);
+      }
+    );
+    for (const ipRange of ipRanges) {
+      const start = this.getIpAsInteger(ipRange.start);
+      const end = this.getIpAsInteger(ipRange.end);
+      for (let j = start; j !== end; j++) {
+        if (!allocInts.includes(j)) {
+          return this.getIntegerAsIp(j);
+        }
+      }
+      if (!allocInts.includes(start) && start === end) {
+        return this.getIntegerAsIp(start);
+      }
+    }
+  }
+
+  /**
+   * Returns if every address within range is in ip list.
+   * @param {Range} range: range object.
+   * @param {string[]} ipList: array of IP addresses.
+   * @return {boolean}
+   */
+  areAllIPsFromRangeInList(ipList: string[], range: Range): boolean {
+    const start = this.getIpAsInteger(range.start);
+    const end = this.getIpAsInteger(range.end);
+    const ipListInts = ipList.map(
+      (ipAddress: string) => {
+        return this.getIpAsInteger(ipAddress);
+      }
+    );
+    for (let i = start; i <= end; i++) {
+      if (!ipList.includes(this.getIntegerAsIp(i)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Indicates whether an IP address is within any range, among a given array of
+   * IP ranges.
+   * @param {string} ip: IP address string.
+   * @param {Range[]} ranges: an array of IP address range objects.
+   * @return {boolean} boolean value.
+   */
+  isIpInAnyRange(ip: string, ranges: Range[]): boolean {
```

Comment:
we have a precedent of using Array<?> notation for expressing array types. please update those everywhere.